fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -536,4 +569,273 @@ private static <T extends ScalarOperator> void collect(ScalarOperator root, Clas
+                ConstantOperator lowerBound =
+                        (ConstantOperator) SqlToScalarOperatorTranslator.translate(range.lowerEndpoint().getKeys().get(0));
+                BinaryPredicateOperator lowerPredicate = new BinaryPredicateOperator(
+                        BinaryPredicateOperator.BinaryType.GE, partitionScalar, lowerBound);
+
+                ConstantOperator upperBound =
+                        (ConstantOperator) SqlToScalarOperatorTranslator.translate(range.upperEndpoint().getKeys().get(0));
+                BinaryPredicateOperator upperPredicate = new BinaryPredicateOperator(
+                        BinaryPredicateOperator.BinaryType.LT, partitionScalar, upperBound);
+
+                CompoundPredicateOperator andPredicate = new CompoundPredicateOperator(
+                        CompoundPredicateOperator.CompoundType.AND, lowerPredicate, upperPredicate);
+                rangeParts.add(andPredicate);
+            } else if (range.hasUpperBound()) {
+                ConstantOperator upperBound =
+                        (ConstantOperator) SqlToScalarOperatorTranslator.translate(range.upperEndpoint().getKeys().get(0));
+                BinaryPredicateOperator upperPredicate = new BinaryPredicateOperator(
+                        BinaryPredicateOperator.BinaryType.LT, partitionScalar, upperBound);
+                rangeParts.add(upperPredicate);
+            } else if (range.hasLowerBound()) {
+                ConstantOperator lowerBound =
+                        (ConstantOperator) SqlToScalarOperatorTranslator.translate(range.lowerEndpoint().getKeys().get(0));
+                BinaryPredicateOperator lowerPredicate = new BinaryPredicateOperator(
+                        BinaryPredicateOperator.BinaryType.GE, partitionScalar, lowerBound);
+                rangeParts.add(lowerPredicate);
+            } else {
+                LOG.warn("impossible to reach here");
+            }
+        }
+        return rangeParts;
+    }
+
+    public static List<OptExpression> collectScanExprs(OptExpression expression) {
+        List<OptExpression> scanExprs = Lists.newArrayList();
+        OptExpressionVisitor scanCollector = new OptExpressionVisitor<Void, Void>() {
+            @Override
+            public Void visit(OptExpression optExpression, Void context) {
+                for (OptExpression input : optExpression.getInputs()) {
+                    super.visit(input, context);
+                }
+                return super.visit(optExpression, context);
+            }
+
+            @Override
+            public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+                scanExprs.add(optExpression);
+                return null;
+            }
+        };
+        expression.getOp().accept(scanCollector, expression, null);
+        return scanExprs;
+    }
+
+    @VisibleForTesting
+    public static Pair<Table, Column> getPartitionTableAndColumn(Expr partitionExpr,
```

Comment:
I  have seen this function many times, can we  unify this ?

Reply:
use the getPartitionTableAndColumn of MaterializedView